stdlib/asm/crypto/stark/random_coin.masm

```diff
@@ -211,31 +376,60 @@ end
 
 #! Draw deep composition polynomial random coefficients and save them into memory in the region from
 #! `deep_rand_coef_ptr` `compos_coef_ptr + 86 - 1` as `(r1_1, r1_0, r0_1, r0_0)`
+#! The number of equal coefficients is equal to:
+#! (72 + 9) * 4 Felt for the trace.
+#! 8 * 2 Felt for constraint polynomial.
+#! 2 Felt for degree correction.
+#! Total: 172 tuples of type (Felt, Felt)
 #!
 #! Input: [deep_rand_coef_ptr, ...]
 #! Output: [...]
-#! Ccycles: 5775
+#! Ccycles: 992
 export.generate_deep_composition_random_coefficients
-    # Load Seed and prepare for drawing
-    push.1
+
     push.0.0.0.0
-    mem_loadw.SEED_PTR
-    # => [Seed, 1, ptr, ..]
-    
-    # Draw the random coefficients
-    # The number of coefficients is equal to:
-    # (72 + 9) * 4 Felt for the trace.
-    # 8 * 2 Felt for constraint polynomial.
-    # 2 Felt for degree correction.
-    # Total: 172 tuples of type (Felt, Felt)
-    repeat.86
-        exec.generate_next_extfelt_tuple
-        mem_loadw.SEED_PTR
-        # => [Seed, 3, ptr + 1]
+    mem_loadw.SEED_PTR       
+    push.0.0.0.0
+    mem_loadw.SEED_PTR_1      
+    push.0.0.0.0
+    mem_loadw.SEED_PTR_2       
+    # => [R2, R1, C, compos_coef_ptr, ...]
+
+    dup.12
+    add.1
+    mem_storew
+
+    swapw
+    dup.12
+    add.2
+    swap.13
+    mem_storew
+    swapw
+
+    repeat.42
+
+        hperm 
+
+        dup.12
+        add.1
+        mem_storew
+
+        swapw
+        dup.12
+        add.2
+        swap.13
+        mem_storew
+        swapw
     end
```

Comment:
I wonder if we should have a helper procedure which would generate a sequence of random elements of variable length, and then call this procedure from here and from other places. The signature for the procedure could be:
```
#! Input: [dest_ptr, num_tuples, ...]
#! Output: [...]
proc.generate_random_coefficients
```
To simplify things, we can assume that the procedure always outputs tuples in increments of 4 (i.e., 8 field elements) and the calling procedure would ignore w/e is extra.

The inner loop in this procedure could look like so:
```
while.true
  # => [loop_ctr, dest_ptr, x, x, R1, C, R2, ...]
  # where loop_ctr = - num_tuples / 4; we negate the counter so that we can count up to 0

  swapw.3 hperm
  # => [R2, R1, C, loop_ctr, dest_ptr, x, x, ...]

  # save R2 to mem[dest+1]; we use dup.13 here because it takes only 1 cycle
  dup.13 add.1 mem_storew
  # => [R2, R1, C, loop_ctr, dest_ptr, x, x, ...]

  # save R1 to mem[dest]
  swapw dup.13 mem_storew swapw
  # => [R2, R1, C, loop_ctr, dest_ptr, x, x, ...]

  # update destination pointer and loop counter
  swapw.3
  # => [loop_ctr, dest_ptr, x, x, R1, C, R2, ...]

  swap add.2 swap
  # => [loop_ctr, dest_ptr+2, x, x, R1, C, R2, ...]

  add.1 dup
  # => [loop_ctr+1, loop_ctr+1, dest_ptr+2, x, x, R1, C, R2, ...]

  eq.0
end
```
I believe this inner loop should take about 21 cycles - which very similar to the highlighted loop. We could save a couple hundred cycles by using `repeat` instead of `while` - but I'm not sure it is worth it.

Reply:
Correction - the last instruction in the loop body above should be `neq.0`.